### PR TITLE
Purge old DELETEs during live compaction

### PR DESCRIPTION
**Pull request by weggert**
This change is to address Issue 112, 'Old deletes need to be purged
during Compaction'.

It adds the ability to specify a
sirius.log.compaction-max-delete-age-hours configuration property to
cause old Delete events to be purged during live compaction.

The configuration property defaults to a very large value, so by default
Delete events will not be purged.

It also adds an option to WalTool to purge old Deletes, along with
other options for old PUTs.

### CODE REVIEW

**Comment by weggert** [comment]
Pull request https://github.com/Comcast/sirius/pull/113 addresses this issue.

File: src/main/scala/com/comcast/xfinity/sirius/tool/WalTool.scala [hunk 1]
```diff
@@ -317,6 +355,42 @@ object WalTool {
...
+    }
+
+    inWal.foreach {
+      case evt if filterDeletesOlderThan(evt, earliestTimestamp) != None => outWal.writeEntry(evt)
+      case _ =>
+    }
+  }
+
+  private def filterDeletesOlderThan(evt: OrderedEvent, earliestTimestamp: Long): Option[OrderedEvent] = {
+    val request = evt.request
```
Comment (pcleary00) [inline src/main/scala/com/comcast/xfinity/sirius/tool/WalTool.scala]: Don't really need the local variable here, can just do `evt.request match`

File: src/main/scala/com/comcast/xfinity/sirius/api/SiriusConfiguration.scala [hunk 2]
```diff
@@ -195,6 +195,12 @@ object SiriusConfiguration {
   final val COMPACTION_SCHEDULE_MINS = "sirius.log.compaction-schedule-mins"
 
   /**
+   * The maximum age for a Delete in hours that remains after compaction.  Deletes
+   * that are older than this age will be purged during compaction.
```
Comment (clinedome-work) [inline src/main/scala/com/comcast/xfinity/sirius/api/SiriusConfiguration.scala]: Can you add text about the default value here?

**Comment by weggert** [comment]
I added documentation to the purge deletes config item, and removed the unnecessary variable from WalTool.

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactor.scala [hunk 3]
```diff
@@ -15,15 +15,28 @@
...
 import annotation.tailrec
 
 object SegmentedCompactor {
-
   val COMPACTING_SUFFIX = ".compacting"
   val TEMP_SUFFIX = ".temp"
 
+  def apply(siriusConfig: SiriusConfiguration) : SegmentedCompactor = {
+
+    val maxDeleteAgeHours = siriusConfig.getProp(SiriusConfiguration.COMPACTION_MAX_DELETE_AGE_HOURS, Integer.MAX_VALUE) // a really long time by default
```
Comment (clinedome-work) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactor.scala]: Please remove the trailing comment, I think it goes over the 120 char max width. It can either go on the line above or it can just be removed, I think it's fine either way.

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactor.scala [hunk 4]
```diff
@@ -15,15 +15,28 @@
...
+    val maxDeleteAgeMillis = 60L * 60 * 1000 * maxDeleteAgeHours
+
+    new SegmentedCompactor(maxDeleteAgeMillis)
+  }
+}
+
+class SegmentedCompactor(maxDeleteAgeMillis: Long) {
```
Comment (clinedome-work) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactor.scala]: This can probably made at least package private. Outside of tests, the apply() method should be used to instantiate this class.

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactor.scala [hunk 5]
```diff
@@ -92,17 +105,19 @@ object SegmentedCompactor {
...
    *
    * @param keys keys to gc against
    * @param source source segment, to be GC'ed
    * @param dest destination segment, empty, will be populated
    */
   private def compactSegment(keys: Set[String], source: Segment, dest: Segment) {
-    // XXX if we want to remove old deletes from the log...
-    // val currentTime = System.currentTimeMillis
+    val currentTime = System.currentTimeMillis
+    val earliestDelete = if (maxDeleteAgeMillis > currentTime) 0L else currentTime - maxDeleteAgeMillis
```
Comment (clinedome-work) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactor.scala]: This could use a comment. I guess none of the updates are likely to come from the 60's, so we're probably not missing any deletes with this, but it's a little confusing, and I'm not sure it makes this any safer/simpler/easier. We could probably remove the `if` statement and just keep it simply `currentTime - maxDeleteAgeMillis`

File: src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactor.scala [hunk 6]
```diff
@@ -92,17 +105,19 @@ object SegmentedCompactor {
...
+
     source.foreach {
-      // case OrderedEvent(_, timestamp, Delete(key)) if (currentTime - timestamp) > MAX_DELETE_AGE => // nothing
+      case OrderedEvent(_, timestamp, Delete(key)) if timestamp < earliestDelete => // nothing
```
Comment (clinedome-work) [inline src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactor.scala]: I do like using `earliestDelete` so we don't have to see that comparison happen down here.

File: src/test/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactorTest.scala [hunk 7]
```diff
@@ -232,11 +249,34 @@ class SegmentedCompactorTest extends NiceTest with BeforeAndAfterAll {
...
+    writeEvents(second, List(3L, 4L, 5L, 6L), now)
+
+    val siriusConfigWithMaxAge = new SiriusConfiguration()
+    siriusConfigWithMaxAge.setProp(SiriusConfiguration.COMPACTION_MAX_DELETE_AGE_HOURS, 1)
+    val segmentedCompactor = SegmentedCompactor(siriusConfigWithMaxAge)
+    val compactionMap = segmentedCompactor.compactAgainst(second, segments)
+
+    val compacted = makeSegment(compactionMap(first))
+
+    assert("2" === listEvents(compacted))
```
Comment (clinedome-work) [inline src/test/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactorTest.scala]: This looks good.

**Comment by clinedome-work** [comment]
All in all, this looks great. A few minor comments, but with those addressed, I'd be +1. Thanks @weggert !

**Comment by weggert** [comment]
Per the feedback I pushed up changes to make the SegmentedCompactor private, remove the unnecessary comment, and simplify the Delete purge logic in the SegmentedCompactor.

**Comment by clinedome-work** [comment]
Awesome. Looks great. :+1:

File: src/main/scala/com/comcast/xfinity/sirius/tool/WalTool.scala [hunk 8]
```diff
@@ -81,12 +81,24 @@ object WalTool {
     Console.err.println("       Same as key-filter, except the resulting UberStore contains all")
     Console.err.println("       OrderedEvents not matching regexp")
     Console.err.println()
+    Console.err.println("   key-filter-deletes-not-older-then <earliestMillis> <inWalDir> <outWalDir>")
```
Comment (tbarker9comcast) [inline src/main/scala/com/comcast/xfinity/sirius/tool/WalTool.scala]: `s/then/than`

File: src/test/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactorTest.scala [hunk 9]
```diff
@@ -232,11 +249,34 @@ class SegmentedCompactorTest extends NiceTest with BeforeAndAfterAll {
     }
   }
 
+  it("should remove all delete events in appropriate segments that are older than COMPACTION_MAX_DELETE_AGE_HOURS") {
```
Comment (tbarker9comcast) [inline src/test/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactorTest.scala]: Since we are considering the corner case in the code, can you create a test when the age is greater than the current time (ie nothing is deleted)?

**Comment by tbarker9comcast** [comment]
Had some minor comments, but once they are addressed this has my :+1:. This looks great to me.

**Comment by weggert** [comment]
I fixed the typo in the WalTool documentation, and I added a new test for the SegmentedCompactor Delete purge logic to make sure it works as expected when the max-time is bigger than the current time.

There is also a minor change for when getting the property to change siriusConfig.getProp(...) to siriusConfig.getInt(...) to avoid ClassCastExceptions if the property value is set with a Long.

**Comment by weggert** [comment]
Thanks for the reviews.  :-)  I squashed the updates into the original commit

**Comment by clinedome-work** [comment]
:+1:
